### PR TITLE
fix: respect user configuration for work with status codes

### LINKED ISSUE #708: Refactor handling of `ignored_http_status_codes` and `SessionError`
- Technical debt introduced in https://github.com/apify/crawlee-python/pull/167
- Should probably be ported over to JS afterwards (v4?)

The reasoning behind the change was that some errors (such as 401) are automatically considered `SessionError` and http-based crawlers don't take `ignore_htttp_error_status_codes` into account for them. While uncommon, it should be possible to explicitly ignore any status code.

### PR DESCRIPTION
**Pull request by Mantisus**
### Description

Fixes HTTP status code configuration behavior:

- `additional_http_error_status_codes`: Properly triggers errors and retries when specified status codes are encountered
- `ignore_http_error_status_codes`: Correctly treats specified error status codes as successful responses

### Issues

- Closes: #756 
- closes #708

### CODE REVIEW

**Comment by Mantisus** [comment]
I have already discussed this with @janbuchar. This PR should not introduce any breaking changes. If `additional_http_error_status_codes` and `ignore_http_error_status_codes` are not specified, the previous behavior is preserved.

Status codes were previously discussed in PR: #498 

However, we have differences with the JS version of `crawlee`:

Python:
- 400-499 codes cause errors without retries

JS:  
- 400-499 codes are treated as successful
- Exceptions:
 - 401, 403, 429 - cause errors with retries
 - 406 - causes error without retries and ignores `ignoreHttpErrorStatusCodes` parameter

File: src/crawlee/sessions/_session.py [hunk 1]
```diff
@@ -193,17 +193,22 @@ def is_blocked_status_code(
         self,
         *,
         status_code: int,
-        additional_blocked_status_codes: list[int] | None = None,
+        additional_blocked_status_codes: set[int] | None = None,
+        ignore_http_error_status_codes: set[int] | None = None,
```
Comment (janbuchar) [inline src/crawlee/sessions/_session.py:197]: Is there even a good reason for this method to be on the `Session` class? Maybe we could move it to the `BasicCrawler` instead.

File: src/crawlee/abstract_http_crawler/_abstract_http_crawler.py [hunk 2]
```diff
@@ -204,13 +204,10 @@ async def _handle_blocked_request(
...
         if self._retry_on_blocked:
             status_code = context.http_response.status_code
-
-            # TODO: refactor to avoid private member access
-            # https://github.com/apify/crawlee-python/issues/708
-            if (
-                context.session
-                and status_code not in self._http_client._ignore_http_error_status_codes  # noqa: SLF001
-                and context.session.is_blocked_status_code(status_code=status_code)
+            if context.session and context.session.is_blocked_status_code(
```
Comment (janbuchar) [inline src/crawlee/abstract_http_crawler/_abstract_http_crawler.py:207]: Did you check the equivalent method in `PlaywrightCrawler` as well? Any chance we could deduplicate this?

**Review by janbuchar — COMMENTED** [review]
Two questions, otherwise LGTM... much better!

**Review by vdusek — APPROVED** [review]
Just one last naming thing, otherwise LGTM.

File: src/crawlee/http_clients/_base.py [hunk 3]
```diff
@@ -150,8 +150,19 @@ def _raise_for_error_status_code(
...
-            if include_error:
-                raise HttpStatusCodeError('Error status code (user-configured) returned.', status_code)
+        if include_error:
+            raise HttpStatusCodeError('Error status code (user-configured) returned.', status_code)
 
+        if is_status_code_client_error(status_code) and not exclude_error:
+            raise HttpClientStatusCodeError('Client error status code returned', status_code)
+
+        if is_status_code_server_error(status_code) and not exclude_error:
             raise HttpStatusCodeError('Error status code returned', status_code)
```
[the remark this answers is not in the document]
Comment (Mantisus) [reply src/crawlee/http_clients/_base.py]: Updated

File: src/crawlee/sessions/_session.py (hunk 1 above)
Comment (Mantisus) [reply src/crawlee/sessions/_session.py:197]: I'm not sure if this is a compelling reason. But if our public API allows configuring errors at the Session level through:

```python
Session(blocked_status_codes=[...])
```

and

```python
SessionPool(create_session_settings={'blocked_status_codes': [...]})
```

then I believe it makes sense to have this method at the Session class level.

File: src/crawlee/abstract_http_crawler/_abstract_http_crawler.py (hunk 2 above)
Comment (Mantisus) [reply src/crawlee/abstract_http_crawler/_abstract_http_crawler.py]: This is possible after refactoring as @Pijukatel did with `AbstractHttpParser`. So for now, I've only added a general method for dealing with status codes.

File: src/crawlee/sessions/_session.py (hunk 1 above)
Comment (janbuchar) [reply src/crawlee/sessions/_session.py:197]: Huh, yeah, I'm not sure if enabling this in `SessionPool` is in fact necessary, but let's not change it in this PR :smile:

File: src/crawlee/sessions/_session.py [hunk 4]
```diff
@@ -193,17 +193,22 @@ def is_blocked_status_code(
...
+            ignore_http_error_status_codes: Optional status codes to allow suppression of
+            codes from `blocked_status_codes`.
 
         Returns:
             True if the session should be retired, False otherwise.
         """
-        blocked_status_codes = self._blocked_status_codes + (additional_blocked_status_codes or [])
+        blocked_status_codes = self._blocked_status_codes | (additional_blocked_status_codes or set())
+
+        blocked_status_codes = blocked_status_codes - (ignore_http_error_status_codes or set())
```
Comment (janbuchar) [inline src/crawlee/sessions/_session.py:212]: Not sure if I'm reading this right, but it seems to me that `ignore_http_error_status_codes` takes precedence over `additional_blocked_status_codes` here, but in `BaseHttpClient`, the `additional_http_error_status_codes` option takes precedence over `ignore_http_error_status_codes`.

This is an edge case for sure, but still, I'd like to keep things consistent.

**Comment by janbuchar** [comment]
> > However, we have differences with the JS version of crawlee:
> > Python:
> > 400-499 codes cause errors without retries
> > JS:
> > 400-499 codes are treated as successful
> > Exceptions:
> > 401, 403, 429 - cause errors with retries
> > 406 - causes error without retries and ignores ignoreHttpErrorStatusCodes parameter
> 
> Any ideas how we want to align it? I believe we do not have issue for that, correct? cc @janbuchar

I believe https://github.com/apify/crawlee/issues/812 tracks this. I also think that what we do in this PR is the correct way to handle things. The "success on 4xx" in JS is a crutch that prevented retries on such errors, and we can do that more cleanly now (in both implementations).

File: src/crawlee/abstract_http_crawler/_abstract_http_crawler.py (hunk 2 above)
Comment (janbuchar) [reply src/crawlee/abstract_http_crawler/_abstract_http_crawler.py]: Nice, thank you!

File: src/crawlee/sessions/_session.py (hunk 4 above)
Comment (Mantisus) [reply src/crawlee/sessions/_session.py]: You're right.

Updated the code to make `additional_http_error_status_codes` take precedence in both cases, which matches the behavior in the JS version

File: src/crawlee/basic_crawler/_basic_crawler.py [hunk 5]
```diff
@@ -1074,3 +1073,10 @@ async def __run_task_function(self) -> None:
 
     async def __run_request_handler(self, context: BasicCrawlingContext) -> None:
         await self._context_pipeline(context, self.router)
+
+    def _is_blocked_status_code(self, session: Session | None, status_code: int) -> bool:
```
Comment (janbuchar) [inline src/crawlee/basic_crawler/_basic_crawler.py:1077]: This one deserves a docblock.

**Review by janbuchar — COMMENTED** [review]
One more thing - it looks like there are no tests for this behavior - could you add some? test_http_crawler would be an appropriate place I guess.

File: tests/unit/http_crawler/test_http_crawler.py [hunk 6]
```diff
@@ -136,19 +136,48 @@ async def test_handles_redirects(
     assert server['html_endpoint'].called
 
 
+@pytest.mark.parametrize(
+    ('additional_http_error_status_codes', 'ignore_http_error_status_codes', 'expected_number_error'),
+    [
+        ([], [], 1),
+        ([403], [], 3),
+        ([], [403], 0),
+        ([403], [403], 3),
```
Comment (janbuchar) [inline tests/unit/http_crawler/test_http_crawler.py:145]: So this is not considered a `SessionError`, not before this PR and not after?

Comment (Mantisus) [reply tests/unit/http_crawler/test_http_crawler.py:145]: That's right

**Review by janbuchar — APPROVED** [review]

**Pull request merged**